client/post-editor/post-editor.jsx

```diff
@@ -91,6 +91,27 @@ import EditorGutenbergBlocksWarningDialog from './editor-gutenberg-blocks-warnin
  */
 import './style.scss';
 
+/*
+ * Throttle and debounce and callback. Used for autosave.
+ * - run the callback at most every `throttleMs` milliseconds (don't autosave too often)
+ * - wait for at least `debounceMs` before first call (don't autosave while still typing)
+ */
+function throttleAndDebounce( fn, throttleMs, debounceMs ) {
+	const throttled = throttle( fn, throttleMs );
+	const debounced = debounce( throttled, debounceMs );
+
+	const throttledAndDebounced = function() {
+		return debounced.apply( this, arguments );
```

Comment:
I'm having some trouble grokking this line. Won't the `this` here always have the same value, namely that of the module top level? Is there any advantage to this syntax vs. `.call()` or `(...args) => debounced(...args)`?

Reply:
This is also a bit confusing to me, maybe even `null` will work as well and it will be clearer that we don’t really care about the context when calling `debounced`. On some level I understand the need for consistency – this is the same as calling `debounced(…)`, just a bit more confusing to most people :)

Reply:
This way of calling preserves the `this` binding of the wrapped function call:
```js
// trivial wrapper creator
function wrap( fn ) {
  return function() {
    return fn.apply( this, arguments );
  }
}

// class that uses `wrap` to create methods
class A {
  this.wrappedSave = wrap( this.save );

  save() {
    // this would be undefined here if `wrap` didn't use `apply`
    console.log( this );
  }

  onSave() {
    // calls the function produced by `wrap`
    this.wrappedSave();
  }
}
```
All helpers that "decorate" a passed function and return a new one should do that. See, for example, the [Lodash `_.before` implementation](https://github.com/lodash/lodash/blob/4.17.13/lodash.js#L10050).

Reply:
Right. The issue is that the `this` being passed along is not the same. The function being called doesn't get the current value of `this`, but rather the current scope you're calling from. So the wrapped function would get `wrap`'s scope (the top level) as its `this`, rather than the value of `this` as `wrap` sees it (which would be the class scope).

So then you need to resort to an arcane syntax when you want something to get the same `this` you got. Oh, JavaScript...